Improved handling of jsonschema

radix0000 opened the pull request:
Handle urn references and allOf conditional in schema_dict_fields_generator

odscjames commented:
Are `resolve_ref` and `process_items` intended to be called outside the module? If so, can we have some doc strings about expected input and output and if not can they be marked private with `_`?

It would be good to add some tests for this too - I realise this doesn't have any existing tests and I can take the work to add tests to existing code you can build on if you want?

odscjames commented:
I'm not sure we need "requirement*.txt" especially as the contents of the dev one contradict setup.py? Can we just leave them out and put any needed restrictions in setup.py and do "pip install setup-e .[dev]" in the workflow?

odscjames commented on tests/test_orgids_prefixes.py:
```diff
@@ -0,0 +1,10 @@
+from libcove2.common import get_orgids_prefixes
+
+
+def test_get_orgids_prefixes():
```
Might be worth a note that this test will fail if internet access or org id server  is down but we can do that later.

odscjames commented:
> which I assume are all the externally called functions. Note there are some others which aren't marked as private, but assume probably are, and are just helpers for ones listed above 

It's a fair point it may not be clear at the moment around existing ones, I'm happy to take that as a issue to clear up later.

odscjames commented on libcove2/common.py:
```diff
@@ -12,32 +12,116 @@
+    return items_schema_dicts
+
+
+def schema_dict_fields_generator(schema_dict, registry=None, defs=None):
```
Could we have a note in docstring of what type of var registry and defs are expected to be?

odscjames replied:
I'm still slightly confused what exactly registry should be but I'm happy to merge this in and then I'll send you a docs PR and we can fine tune

radix0000 commented:
Okay, have removed "requirement*.txt" files, added a comment and a docstring.

radix0000 commented:
Actually bumped the actions/cache to v4 in workflows as well, don't know if people saw that email about v1-2 being closed down soon.

Commit:
common.py: Improved handling of jsonschema
Handle urn references and allOf conditional in
schema_dict_fields_generator (plus added tests)

jpmckinney commented on libcove2/common.py:
```diff
@@ -12,32 +12,132 @@
 
 
-def schema_dict_fields_generator(schema_dict):
+def _resolve_ref(value, defs, registry=None):
```
From the code I've seen, schema_dict_fields_generator is passed a schema on which jsonref.replace_refs (or equivalent) has already been called. jsonref's default loader doesn't support URNs, but this support can be added by using a custom [loader](https://jsonref.readthedocs.io/en/latest/#jsonref.replace_refs) instead of the default [jsonloader](https://jsonref.readthedocs.io/en/latest/#jsonref.jsonloader). It would be cleaner to resolve the refs outside schema_dict_fields_generator, rather than giving schema_dict_fields_generator the responsibility for cleaning up the refs left unresolved by the default loader.

odscjames replied:
Me and @radix0000 have had a few chats about this before and we just discussed it again. 0.2 release of this library has been made, so it is part of functionality for now (and we have a release that's dependent on this).  

But we will research some questions we have about schema wrangling before we add more like this to this library. I'd also like to get to a place where we can pass a fully de-referenced schema - not just for schema_dict_fields_generator, but for all the other functions CoVE has that might use a schema. If we can sort that out we'll unpick this work by clearly documenting expectations and marking the relevant parameters deprecated before removing them later.

jpmckinney replied:
FYI, for the OCDS Data Review Tool, I’m switching to using get_schema_fields in OCDS Kit for all schema analysis. This originally was only for the mapping sheet. https://github.com/open-contracting/ocdskit/blob/c3f341b1dd3d941a3ac4443203d7e271509fae22/ocdskit/schema.py#L66

It takes a deferenced schema and extracts all metadata (deprecated, required etc) in a single pass. There are tests (test_regression) to check that it can produce the same results as libcove with equal or better running time. As far as I know, it supports BODS except for URN support (but that support can be added to the jsonref loader).

The tests in test_regression make a few notes of bugs in libcove (missing support for oneOf in the deprecated and required ID functions, etc.).

jpmckinney commented on tests/test_schema_dict_fields_generator.py:
```diff
@@ -0,0 +1,33 @@
+
+    schema_fields = set(schema_dict_fields_generator(schema_0_3))
+
+    assert len(schema_fields) == 130
```
There must be a bug in the code. There should be 138 fields. These aren't returned:

```python
{'/annotations/url', '/annotations/createdBy', '/annotations/statementPointerTarget', '/annotations/createdBy/uri', '/annotations/creationDate', '/annotations/description', '/annotations/createdBy/name', '/annotations/transformedContent'}
```

jpmckinney replied:
Ah, the bug is that `annotations/items` sets both `properties` and `oneOf`, but `_process_items` treats these as alternatives (`elif`).

radix0000 replied:
@jpmckinney Thanks for spotting this out. There is a new PR to fix it: https://github.com/OpenDataServices/lib-cove-2/pull/14 and more tests have been added.
